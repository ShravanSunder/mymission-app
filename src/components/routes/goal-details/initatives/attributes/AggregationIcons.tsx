import { css } from '@emotion/react';
import { faCalendarAlt, faCalendarDay, faCalendarWeek } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React, { FC } from 'react';
import tw from 'twin.macro';

import { RecurrenceRepetitionAggregation } from './core/recurrence.types';

import { ICommonProps } from '~~/components/common/ICommonProps';

export const AggregationIcons: FC<{ period: RecurrenceRepetitionAggregation } & ICommonProps> = (props) => {
   if (props.period === RecurrenceRepetitionAggregation.PerDay) {
      return <FontAwesomeIcon fixedWidth={true} className={props.className} css={css(tw`w-full h-full`)} icon={faCalendarDay} />;
   } else if (props.period === RecurrenceRepetitionAggregation.PerWeek) {
      return <FontAwesomeIcon fixedWidth={true} className={props.className} css={css(tw`w-full h-full`)} icon={faCalendarWeek} />;
   } else if (props.period === RecurrenceRepetitionAggregation.PerMonth) {
      return <FontAwesomeIcon fixedWidth={true} className={props.className} css={css(tw`w-full h-full`)} icon={faCalendarAlt} />;
   } else {
      return null;
   }
};
