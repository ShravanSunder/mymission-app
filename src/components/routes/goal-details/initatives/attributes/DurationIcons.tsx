import { faCalendarAlt, faCalendarPlus, faCalendarWeek, faThLarge } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { EventNote } from '@material-ui/icons';
import { FC } from 'react';
import { muiIconCss } from '~~/helpers/muiIconCss';
import { RecurrenceDurationType } from './core/recurrence.types';

export const DurationIcons: FC<{ duration: RecurrenceDurationType }> = (props) => {
   if (props.duration === RecurrenceDurationType.Weekly) {
      return <FontAwesomeIcon fixedWidth={true} className="w-full h-full" icon={faCalendarWeek} />;
   } else if (props.duration === RecurrenceDurationType.SpecificDaysOfWeek) {
      return <FontAwesomeIcon fixedWidth={true} className="w-full h-full" icon={faCalendarPlus} />;
   } else if (props.duration === RecurrenceDurationType.Monthly) {
      return <FontAwesomeIcon fixedWidth={true} className="w-full h-full" icon={faCalendarAlt} />;
   } else if (props.duration === RecurrenceDurationType.Quarterly) {
      return <FontAwesomeIcon fixedWidth={true} className="w-full h-full" icon={faThLarge} />;
   } else {
      return <EventNote css={muiIconCss}></EventNote>;
   }
};
