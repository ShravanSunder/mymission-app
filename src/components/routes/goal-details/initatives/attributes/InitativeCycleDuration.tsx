import { Typography } from '@material-ui/core';
import React, { useMemo } from 'react';
import { TwemojiInline } from '~~/components/common/Twemoji';
import { RecurrenceTypes, recurrenceToDisplayString, recurrenceToNumberOfDaysMap, DaysOfWeek } from './scheduleDefinitions';

export interface IInitativeCycleDurationProps {
   recurrenceType?: RecurrenceTypes;
}
export const InitativeCycleDurationSummary: React.FC<IInitativeCycleDurationProps> = (props) => {
   const summary = useMemo(() => {
      // if (recurrenceToNumberOfDaysMap.has(props.recurrenceType)) {
      //    return recurrenceToDisplayString(props.recurrenceType, props.recurrenceSchedule as number);
      // } else {
      //    // Todo:finish me
      //    return '';
      // }
      return '';
   }, [props.recurrenceType]);

   return (
      <div
         className="w-full select-none grid grid-row-3 "
         css={{
            gridTemplateColumns: 'auto 1fr 1fr ',
         }}>
         <div className="w-full rounded-full">
            <TwemojiInline text="ðŸ“…"></TwemojiInline>
         </div>
         <Typography className="pl-2 capitalize">
            <strong>Days</strong>
         </Typography>
         <Typography className="pr-2 text-right justify-self-end">{summary}</Typography>
      </div>
   );
};

export const InitativeCycle: React.FC<IInitativeCycleDurationProps> = (props) => {
   return (
      <div className="w-full overflow-hidden overflow-y-auto grid grid-cols-1 max-h-80">
         <div className="">{props.recurrenceType}</div>
      </div>
   );
};
