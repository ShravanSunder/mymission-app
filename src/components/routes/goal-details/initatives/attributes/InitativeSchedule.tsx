import { Accordion, AccordionDetails, AccordionSummary, IconButton } from '@material-ui/core';
import { useState, FC } from 'react';
import tw from 'twin.macro';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import { DaysOfWeek } from './scheduleDefinitions';
import { RecurrenceAggregationPeriods, RecurrenceDurationTypes } from './recurrenceDefinitions';
import { RecurrenceDuration } from './RecurrenceDuration';
import { ScheduleAccordionSummary } from './ScheduleAccordionSummary';
import { useRecurrenceSummary } from './useRecurrenceSummary';
import { RecurrenceAggregationPeriod } from './RecurrenceAggregationPeriod';
import { useIntl } from 'react-intl';
import { useAggregationText } from './useAggregationText';

const tempColorIcons = 'bg-gray-200';

export const InitativeSchedule: FC = (props) => {
   const { formatMessage } = useIntl();
   const getAggregationText = useAggregationText();

   const [recurrenceAggregationPeriod, setRecurrenceAggregationPeriod] = useState(RecurrenceAggregationPeriods.PerDay);

   const aggregationName = formatMessage({ defaultMessage: 'Habit counting' });
   const aggregationValue = getAggregationText(recurrenceAggregationPeriod);

   const aggregateAccordion = (
      <Accordion>
         <AccordionSummary expandIcon={<ExpandMoreIcon />} id="panel2a-header">
            <ScheduleAccordionSummary summaryName={aggregationName} summaryValue={aggregationValue.primary}></ScheduleAccordionSummary>
         </AccordionSummary>
         <AccordionDetails>
            <RecurrenceAggregationPeriod
               aggregationPeriod={recurrenceAggregationPeriod}
               setAggregationPeriod={setRecurrenceAggregationPeriod}></RecurrenceAggregationPeriod>
         </AccordionDetails>
      </Accordion>
   );

   const [recurrenceDuration, setRecurrence] = useState<RecurrenceDurationTypes>(RecurrenceDurationTypes.Monthly);
   const [recurrenceTarget, setRecurrenceSchedule] = useState<number | DaysOfWeek[]>(5);
   const durationSummaryValue = useRecurrenceSummary(recurrenceAggregationPeriod, recurrenceDuration, recurrenceTarget);

   const durationAccordion = (
      <Accordion>
         <AccordionSummary expandIcon={<ExpandMoreIcon />} id="panel2a-header"></AccordionSummary>
         <AccordionDetails>
            <RecurrenceDuration
               recurrenceSchedule={recurrenceTarget}
               recurrenceType={recurrenceDuration}
               setRecurrence={setRecurrence}
               setRecurrenceSchedule={setRecurrenceSchedule}></RecurrenceDuration>
         </AccordionDetails>
      </Accordion>
   );
   return (
      <div className="w-full">
         {aggregateAccordion}
         {durationAccordion}
      </div>
   );
};
