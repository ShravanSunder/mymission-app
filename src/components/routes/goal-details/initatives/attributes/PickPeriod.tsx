import { useEffect, ReactNode, MouseEvent, FC } from 'react';
import { daysToRecurrenceTypeMap } from './recurrenceDefinitions';
import { css } from '@emotion/react';
import tw from 'twin.macro';
import { IconButton, Typography } from '@material-ui/core';
import { IRecurrenceDurationProps } from './RecurrenceDuration';

export const PickPeriod: FC<IRecurrenceDurationProps> = (props) => {
   /**
    * todo: replace colors
    */
   const tempColorSelectedDay = 'bg-gray-200';

   const handleChange = (event: MouseEvent<HTMLElement> | null, newValue: number | null) => {
      if (newValue) {
         props.target.subject$.next(newValue);
      }
   };

   let days: ReactNode[] | null = null;

   const availableDays = daysToRecurrenceTypeMap.get(props.durationType.state);
   if (availableDays != null && availableDays > 0) {
      const resultDays: ReactNode[] = [];

      for (let i = 1; i <= availableDays; i++) {
         let selectStyle = css();
         if (props.target.state === i) {
            selectStyle = css(tw`${tempColorSelectedDay} shadow-sm`);
         }

         resultDays.push(
            <div css={selectStyle} key={i} className="rounded-full w-11 h-11">
               <IconButton className="" value={i} onClick={() => handleChange(null, i)}>
                  <Typography className="w-5 h-5" variant="subtitle2">
                     {i}
                  </Typography>
               </IconButton>
            </div>
         );
      }

      days = resultDays;
   }

   return <div className="flex flex-wrap content-center w-full h-full place-self-center justify-items-start">{days}</div>;
};
