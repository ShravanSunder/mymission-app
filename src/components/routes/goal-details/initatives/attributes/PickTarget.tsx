import { css } from '@emotion/react';
import { IconButton, Typography } from '@material-ui/core';
import { FC, MouseEvent, ReactNode, useMemo } from 'react';
import tw from 'twin.macro';
import { daysToRecurrenceTypeMap } from './core/recurrence.types';
import { DaysOfWeek } from './core/schedule.types';
import { IRecurrenceGoalProps } from './RecurrenceGoal';

interface ITargetItemProps {
   target: number | DaysOfWeek[];
   index: number;
   handleChange: (event: MouseEvent<HTMLElement> | null, newValue: number | null) => void;
}

const TargetItem: FC<ITargetItemProps> = (props) => {
   const tempColorSelectedDay = 'bg-gray-200';
   const index = props.index;

   return useMemo(() => {
      let selectStyle = css();
      if (props.target === index) {
         selectStyle = css(tw`${tempColorSelectedDay} shadow-sm`);
      }

      return (
         <div css={selectStyle} key={index} className="rounded-full w-11 h-11">
            <IconButton className="" value={index} onClick={() => props.handleChange(null, index)}>
               <Typography className="w-5 h-5" variant="subtitle2">
                  {index}
               </Typography>
            </IconButton>
         </div>
      );
   }, [props, index]);
};

export const PickTarget: FC<IRecurrenceGoalProps> = (props) => {
   /**
    * todo: replace colors
    */
   const tempColorSelectedDay = 'bg-gray-200';

   const handleChange = (event: MouseEvent<HTMLElement> | null, newValue: number | null) => {
      if (newValue) {
         props.target.push(newValue);
      }
   };

   let days: ReactNode[] | null = null;
   const availableDays = daysToRecurrenceTypeMap.get(props.durationType.value);
   if (availableDays != null && availableDays > 0) {
      const resultDays: ReactNode[] = [];
      for (let i = 1; i <= availableDays; i++) {
         const target = <TargetItem target={props.target.value} key={i} index={i} handleChange={handleChange} />;
         resultDays.push(target);
      }
      days = resultDays;
   }

   return <div className="flex flex-wrap content-center place-self-center justify-items-start">{days}</div>;
};
