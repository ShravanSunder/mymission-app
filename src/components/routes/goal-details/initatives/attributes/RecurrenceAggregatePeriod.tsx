import { useMemo } from 'react';

import { DaysOfWeek } from './scheduleDefinitions';
import { RecurrenceAggregationPeriods, RecurrenceDurationTypes, daysToRecurrenceTypeMap } from './recurrenceDefinitions';

export const tempColorSelectedDay = 'bg-gray-200';

/**
 * @see IRecurrenceAggregatePeriodProps
 */
export interface IRecurrenceAggregatePeriodSummaryProps {
   aggregationPeriod: RecurrenceAggregationPeriods;
}

export interface IRecurrenceAggregatePeriodProps {
   /**
    * The type of repetition.  ie Days per week, days per month
    */
   recurrenceType: RecurrenceDurationTypes;
   /**
    * Number: Number of times per repetition.
    * Days of Week:  When repetition type is SpecificDaysOfWeek, it can be an DaysOfWeek[]
    */
   RecurrenceAggregatePeriod: number | DaysOfWeek[];

   setRecurrence: React.Dispatch<React.SetStateAction<RecurrenceDurationTypes>>;
   setRecurrenceAggregatePeriod: React.Dispatch<React.SetStateAction<number | DaysOfWeek[]>>;
}

export const RecurrenceAggregatePeriod: React.FC<IRecurrenceAggregatePeriodProps> = (props) => {
   return (
      <div className="w-full overflow-hidden overflow-y-auto grid grid-cols-1 max-h-80">
         <div className="">{props.recurrenceType}</div>
      </div>
   );
};
