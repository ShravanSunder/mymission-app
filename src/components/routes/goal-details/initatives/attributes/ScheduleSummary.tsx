import { Typography } from '@material-ui/core';
import { FC, ReactNode } from 'react';
import { useIntl } from 'react-intl';

import { TwemojiInlineLazy } from '~~/components/common/TwemojiLazy';

interface IScheduleAccordionSummary {
   icon: string;
   summaryName: string;
   summaryValue: string | ReactNode;
}

export const ScheduleSummary: FC<IScheduleAccordionSummary> = (props) => {
   // for test
   // const { formatMessage } = useIntl();
   // const data = formatMessage({ defaultMessage: 'sdfsdfdsfdfds {count}' }, { count: 1 });
   // console.log(data);
   return (
      <div
         className="items-center w-full select-none grid grid-row-3"
         css={{
            gridTemplateColumns: 'auto 1fr auto',
         }}>
         <Typography variant="h3" className="w-full rounded-full">
            <TwemojiInlineLazy text={props.icon}></TwemojiInlineLazy>
         </Typography>
         <Typography variant="h3" className="pl-2 capitalize">
            {props.summaryName}
         </Typography>
         <Typography variant="body1" className="pr-2 text-right justify-self-end">
            {props.summaryValue}
         </Typography>
      </div>
   );
};
