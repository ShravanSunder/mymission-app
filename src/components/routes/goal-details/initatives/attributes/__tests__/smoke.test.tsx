import { screen, render } from '@testing-library/react';
import React from 'react';

import { ScheduleSummary } from '~~/components/routes/goal-details/initatives/attributes/ScheduleSummary';
import { IntlWrapper } from '~~/testing/utils/wrappers';

it('smoke test does render', () => {
   const { asFragment } = render(<ScheduleSummary icon="ðŸ˜€" summaryName="name" summaryValue="sdljfdsjfdskfjs"></ScheduleSummary>, {
      wrapper: IntlWrapper,
   });

   expect(asFragment()).toMatchSnapshot();
   screen.debug();
});
