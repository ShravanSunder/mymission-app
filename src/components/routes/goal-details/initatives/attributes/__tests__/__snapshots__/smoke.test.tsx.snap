// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke test does render 1`] = `
<DocumentFragment>
  <div
    class="items-center w-full select-none grid grid-row-3 css-u5ueop-ScheduleSummary"
  >
    <h3
      class="MuiTypography-root MuiTypography-h3 w-full rounded-full css-gepadz-MuiTypography-root"
    >
      <div />
    </h3>
    <h3
      class="MuiTypography-root MuiTypography-h3 pl-2 capitalize css-gepadz-MuiTypography-root"
    >
      name
    </h3>
    <p
      class="MuiTypography-root MuiTypography-body1 pr-2 text-right justify-self-end css-ahj2mt-MuiTypography-root"
    >
      sdljfdsjfdskfjs
    </p>
  </div>
</DocumentFragment>
`;
